runestone/fitb/fitb.py

```diff
@@ -175,6 +176,15 @@ def run(self):
 
         self.updateContent()
 
+        # Process dynamic problem content.
+        dynamic = self.options.get("dynamic")
+        if dynamic:
+            # Make sure we're server-side.
+            if not env.config.runestone_server_side_grading:
```

Comment:
rats ☹️ only server side....

Reply:
That's always my preference :). Of course, this is also how WebWork operates. Especially if we want to do symbolic math, linear algebra, etc., the code base for Python / server-side programs is more mature than JS.

Reply:
I don't think we are in competition with webwork.  There are things we can emulate from WebWork and there are things we can learn.  My preference is always to push as much of the work as possible to the client, and for basic stuff it seems like it would be very easy to make it work in Javascript as well as Python.

That said, I really like the concept, I think we could do a lot with it.

I've not seen multi-line values for `directives.unchanged` before.  That looks a little wonky, I assume the indentation is important there?

Reply:
I'm not trying to compete with Webwork, but to take ideas / design inspiration from a proven system. I agree that client-side is nice to support, and for basic problems should have all the features. However, trying to support both server-side and client-side seems hard, since the two languages (Python and JavaScript) are fundamentally different animals.

I played with multi-line values for the directive. Correct, they must be indented. More restrictively, there must be no blank lines (although a line with only a comment character can work around that). Perhaps we could also have a `===================` then have the code after that, which is think is what ActiveCode uses?

Reply:
...but thinking about this, if the client-side code is still possible using skulpt, correct? So we could potentially send Python to the browser, then use it there to check answers? We'd also have to do template replacement (e.g. `{{=a}}`, but that seems do-able.


...even more, what if all question types that run either client-side or server-side do all answer checking in Python? Currently, fitb has answer checking written in both languages, which is not great...

Reply:
It could work, the sad thing about skulpt is that it was never designed so that you could have skulpt run some python and return a result to Javascript.  The way skulpt and Javascript communicate (if at all) is through the DOM.

I think we need more discussion on this.  The server side stuff has a lot of benefit, but the fact is that most authors do not install a full runestone server, which makes writing stuff that relies on having a full server difficult.

Reply:
I'd be happy to think more. To me, the core design question is: what language should dynamic problems be written in? Ideally, it would be a language that would execute in the same way on both client and server. Node.js and JavaScript are close, but these don't easily play with our Python-based server. If it is too hard to get Skulpt to talk with JS, perhaps Transcrypt?